### PR TITLE
LEX-1558 Mapping identity and identity type in lift 3.

### PR DESCRIPTION
**Pull request by ynx**

### CODE REVIEW

File: acquia_lift.module [hunk 1]
```diff
@@ -25,12 +25,16 @@ function acquia_lift_help($route_name) {
...
     return;
   }
 
-  // Populate Html head.
-  \Drupal::service('acquia_lift.service.context.page_context')->populateHtmlHead($page['#attached']['html_head']);
+  $page_context = \Drupal::service('acquia_lift.service.context.page_context');
+
+  // Populate contexts.
+  $path_context->populate($page);
+  $page_context->populate($page);
```
Comment (ynx) [inline acquia_lift.module:37]: The contexts are no longer limited to populating the HTML head (for meta tags), they can now populate to the whole page's config array. One immediate use here is that the path context now populates $page['#cache']['context'] to guide a precise caching strategy; while caching as much as possible, the page recalculates as certain conditions are met. More below.

File: config/install/acquia_lift.settings.yml [hunk 2]
```diff
@@ -18,7 +18,7 @@ udf_touch_mappings: []
 udf_event_mappings: []
 visibility:
   # Track, only when the page URL falls OUTSIDE OF these path patterns.
-  path_patterns: "/admin\n/admin/*\n/batch\n/node/add*\n/node/*/*\n/user/*/*\n/block/*"
+  path_patterns: "/admin\n/admin/*\n/batch\n/node/add*\n/node/*/*\n/user/*\n/block/*"
```
Comment (ynx) [inline config/install/acquia_lift.settings.yml:21]: Not exactly related, but fixing an old bug.

File: src/Form/AdminSettingsForm.php [hunk 3]
```diff
@@ -188,15 +188,15 @@ private function buildIdentityForm() {
     $form['identity_parameter'] = [
       '#type' => 'textfield',
       '#title' => t('Identity Parameter'),
-      '#description' => t('The query string parameter for specific visitor information, such as an email address or social media username, which is sent to the Acquia Lift service. Example: ?<ins>@identity_parameter_display_value</ins>=jdoe01', [
+      '#description' => t('The URL link parameter for specific visitor information, such as an email address or social media username, which is sent to the Lift Profile Manager. Example using <strong>@identity_parameter_display_value</strong>: ?<strong><ins>@identity_parameter_display_value</ins></strong>=jdoe01', [
```
Comment (ynx) [inline src/Form/AdminSettingsForm.php:191]: 3 identity descriptions are updated as finalized in the ticket's discussion.

File: src/Form/AdminSettingsForm.php [hunk 4]
```diff
@@ -211,14 +211,19 @@ private function buildIdentityForm() {
...
         '@identity_type_parameter_display_value' => $identity_type_parameter_display_value,
         '@default_identity_type_display_value' => $default_identity_type_display_value,
       ]),
       '#default_value' => $identity_settings['default_identity_type'],
       '#placeholder' => SettingsHelper::DEFAULT_IDENTITY_TYPE_DEFAULT,
+      '#states' => [
+        'visible' => [
+          ':input[name="identity[identity_parameter]"]' => ['!value' => ''],
+        ],
+      ],
```
Comment (ynx) [inline src/Form/AdminSettingsForm.php:226]: Adding missing behaviour - "Default identity type parameter" field should hide if there is no value in "Identity parameter".

File: src/Service/Context/BaseContext.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
+<?php
+
+namespace Drupal\acquia_lift\Service\Context;
+
+abstract class BaseContext implements ContextInterface {
```
Comment (ynx) [inline src/Service/Context/BaseContext.php:5]: Base context to provide support for all child contexts such as Page or Path.

File: src/Service/Context/PathContext.php [hunk 6]
```diff
@@ -103,75 +97,85 @@ public function shouldAttach() {
...
-    if (empty($this->identitySettings['identity_parameter'])) {
+    $identity_parameter = $this->identitySettings['identity_parameter'];
+    if (empty($identity_parameter)) {
       return;
     }
 
+    // Set cache contexts. This is done as long as the identity parameter is set.
+    $identity_type_parameter = $this->identitySettings['identity_type_parameter'];
+    $query_names = [$identity_parameter, $identity_type_parameter];
+    $this->setContextCacheByQueryNames($query_names);
```
Comment (ynx) [inline src/Service/Context/PathContext.php:115]: This is the caching strategy. The strategy is either "url.query_args:MY_IDENTITY" or "url.query_args:MY_IDENTITY and url.query_args:MY_IDENTITY_TYPE". This is depending on the presence of just identity parameter, or both identity and identity type parameters.

File: tests/src/Unit/Service/Context/PathContextTest.php [hunk 7]
```diff
@@ -196,82 +197,258 @@ public function testGetIdentity($query_parameter_string, $capture_identity, $do_
...
+   *
+   * @covers ::populate
+   *
+   * @param integer $expect_set_cache
+   * @param array $identity_settings
+   * @param array $expect_cache_context
+   *
+   * @dataProvider providerTestPopulateCache
+   */
+  public function testPopulateCache($identity_settings, $expect_set_cache, $expect_cache_context) {
```
Comment (ynx) [inline tests/src/Unit/Service/Context/PathContextTest.php:326]: Precise caching strategy is important. Too little means the customer site will suffer, any tiny bit of too much then some page will get cached incorrectly, therefore serving Lift the wrong identity / identity type from the cached page of wrong identities. It's extensively tested here.

**Comment by nickveenhof** [comment]
Looking good!

**Comment by ynx** [comment]
Thanks for the review, @nickveenhof!

**Pull request merged**